### ISSUE TITLE
udf: CALLED ON NULL INPUT with untyped null argument not working

### ISSUE THREAD

**Issue by michae2** [cockroachdb/cockroach#88374]
Looks like we're not passing an untyped null argument into the function call in this case, even though the function uses `CALLED ON NULL INPUT`. Here's the repro:

```sql
CREATE TABLE t (a INT, b INT);
INSERT INTO t VALUES (1, NULL);
CREATE FUNCTION g (i INT) RETURNS INT CALLED ON NULL INPUT LANGUAGE SQL AS $$ SELECT a FROM t WHERE b IS NOT DISTINCT FROM i $$;
SELECT g(NULL);
SELECT g(NULL::INT);
```

On PostgreSQL 14.5, both function calls return 1:

```
michae2=# CREATE TABLE t (a INT, b INT);
CREATE TABLE
michae2=# INSERT INTO t VALUES (1, NULL);
INSERT 0 1
michae2=# CREATE FUNCTION g (i INT) RETURNS INT CALLED ON NULL INPUT LANGUAGE SQL AS $$ SELECT a FROM t WHERE b IS NOT DISTINCT FROM i $$;
CREATE FUNCTION
michae2=# SELECT g(NULL);
 g
---
 1
(1 row)

michae2=# SELECT g(NULL::INT);
 g
---
 1
(1 row)
```

On CockroachDB only the typed null returns 1. The untyped null returns null:

```
demo@127.0.0.1:26257/defaultdb> CREATE TABLE t (a INT, b INT);
CREATE TABLE


Time: 11ms total (execution 11ms / network 0ms)

demo@127.0.0.1:26257/defaultdb> INSERT INTO t VALUES (1, NULL);
INSERT 0 1


Time: 9ms total (execution 9ms / network 0ms)

demo@127.0.0.1:26257/defaultdb> CREATE FUNCTION g (i INT) RETURNS INT CALLED ON NULL INPUT LANGUAGE SQL AS $$ SELECT a FROM t WHERE b IS NOT DISTINCT FROM i $$;
CREATE FUNCTION


Time: 29ms total (execution 29ms / network 0ms)

demo@127.0.0.1:26257/defaultdb> SELECT g(NULL);
   g
--------
  NULL
(1 row)


Time: 1ms total (execution 1ms / network 0ms)

demo@127.0.0.1:26257/defaultdb> SELECT g(NULL::INT);
  g
-----
  1
(1 row)


Time: 7ms total (execution 6ms / network 0ms)
```

I would expect both to return 1.

This is on `v22.2.0-alpha.3-421-g77c412dfef` (tip of master).

Jira issue: CRDB-19790

**Comment by mgartner** [comment]
The FuncExpr is converted to NULL on this line: https://github.com/cockroachdb/cockroach/blob/8248dcea94eeb866429c6a1b4ff0d02aa4934423/pkg/sql/sem/tree/type_check.go#L1144

The problem is that we are type-checking an Overload that we haven't fully resolved - we've only resolved the function signature. This is one of the fears I've had since realizing that we populate the same Overload struct from both a FunctionOverload and a FunctionDescriptor - it's nearly impossible throughout the code to reason about what type of Overload you have. To make this safer, we should have separate in-memory representations of FunctionOverload and FunctionDescriptor so that it's impossible to mistake them. I think we should also rename FunctionOverload to make it more clear that it's only partial information about the function - something like FunctionSignature or FunctionStub. Thoughts @chengxiong-ruan @ajwerner?

As for a fix - is it possible to have an Overload populated by a FunctionDescriptor during type-checking, or do we need to include CalledOnNullInput as a field in FunctionOverload?

**Comment by mgartner** [comment]
Friendly ping @chengxiong-ruan. Do you feel strongly about any of the options I mentioned above?

**Comment by chengxiong-ruan** [comment]
@mgartner sorry for missing this. Yeah, `FunctionSignature` sounds good to me. The only annoying thing is that the overloads of `builtin` functions would be both a signature and an overload because we never have descriptors for them. But I think it's legit to have the new `FunctionSignature` interface embedded in overload interface.

> As for a fix - is it possible to have an Overload populated by a FunctionDescriptor during type-checking, or do we need to include CalledOnNullInput as a field in FunctionOverload?

I think 2 options:
(1) as you said let's include `CalledOnNullInput` as a field in `FunctionOverload`, but it seems semantically weird having it as part of the signature.
(2) How about let's just make function resolution returning overloads (not just signatures). We'll need to do some work to make Schema descriptor to have the full overloads populated when schema descriptor is constructed but that's doable. I'm assuming people wouldn't create/replace functions very often, so performance shouldn't be a problem since we'd only use leased schema descriptors. This is a bit more work than option (1)

I'd prefer option (2), let me know if that make sense.

**Comment by mgartner** [comment]
I'd prefer (2) as well, but I wasn't sure if the full descriptor can be available during type-checking. But based on what you said, it sounds like it should be, correct?

**Comment by chengxiong-ruan** [comment]
Correct, the overload would be populated with full function descriptor.

**Comment by mgartner** [comment]
Great. Is this something you want to take on, or would you prefer SQL Queries handles it?

**Comment by chengxiong-ruan** [comment]
Oh yeah, I'd take it. How urgent is this?

**Comment by mgartner** [comment]
It’s not super urgent, but it is an incorrect result so we should fix it soon. Hopefully the fix can be backported too.

**Comment by mgartner** [comment]
@chengxiong-ruan hold off on the proposed fix for now. I think the type-checking logic is incorrect, and fixing it might fix this bug.

This type checking code will make the function result in `NULL` if there are multiple overloads and none of the resolved overloads are `CALLED ON NULL INPUT`. Postgres does not ignore the ambiguity in potential overloads and returns an error:

```
DROP FUNCTION
marcus=# CREATE FUNCTION f(i INT2) RETURNS INT STRICT LANGUAGE SQL AS 'SELECT 2';
CREATE FUNCTION
marcus=# CREATE FUNCTION f(i INT4) RETURNS INT STRICT LANGUAGE SQL AS 'SELECT 2';
CREATE FUNCTION
marcus=# SELECT f(NULL);
ERROR:  42725: function f(unknown) is not unique
LINE 1: SELECT f(NULL);
               ^
HINT:  Could not choose a best candidate function. You might need to add explicit type casts.
LOCATION:  ParseFuncOrColumn, parse_func.c:577
```

I'll take a shot at fixing this.

**Comment by mgartner** [comment]
The plot has thickened. Built-in strict functions with multiple overloads, like `abs`, can be disambiguated with NULL input, and directly return NULL. Furthermore, creating UDF overloads, both `STRICT` and not `STRICT`, with the same name as these built-ins (e.g., `CREATE FUCNTION abs(t TEXT) ...`) does not make them ambiguous in the presence of NULL inputs. For example:

```
marcus=# SELECT proname, proisstrict, prorettype FROM pg_proc WHERE proname = 'abs';
 proname | proisstrict | prorettype
---------+-------------+------------
 abs     | t           |        700
 abs     | t           |        701
 abs     | t           |         20
 abs     | t           |         23
 abs     | t           |         21
 abs     | t           |       1700
(6 rows)

marcus=# SELECT abs(NULL);
 abs
------
 NULL
(1 row)

marcus=# CREATE FUNCTION abs(t TEXT) RETURNS INT STRICT LANGUAGE SQL AS 'SELECT 1';
CREATE FUNCTION

marcus=# SELECT abs(NULL);
 abs
------
 NULL
(1 row)

marcus=# CREATE FUNCTION abs(t CHAR) RETURNS INT LANGUAGE SQL AS 'SELECT 1';
CREATE FUNCTION

marcus=# SELECT abs(NULL);
 abs
------
 NULL
(1 row)
```

I'm wondering if there are other function-resolution rules that are disambiguating these built-ins that are unrelated to strictness.

**Comment by mgartner** [comment]
I believe the `abs` in the example above is disambiguated by [Postgres's function resolution](https://www.postgresql.org/docs/current/typeconv-func.html) rule:

> If any input arguments are unknown, check the type categories accepted at those argument positions by the remaining candidates. At each position, select the string category if any candidate accepts that category. (This bias towards string is appropriate since an unknown-type literal looks like a string.) Otherwise, if all the remaining candidates accept the same type category, select that category; otherwise fail because the correct choice cannot be deduced without more clues. Now discard candidates that do not accept the selected type category. Furthermore, if any candidate accepts a preferred type in that category, discard candidates that accept non-preferred types for that argument. Keep all candidates if none survive these tests. If only one candidate remains, use it; else continue to the next step.

Our logic which disambiguates `STRICT` functions on `NULL` input is an incorrect oversimplification that tries to mimic this behavior. Furthermore, there is nothing special about built-in functions that aids in disambiguation, as I incorrectly hypothesized above. Built-ins follow the same resolution as UDFs, there is no special logic.

#### Here's how the quoted logic above plays out to resolve to a single overload (when no `abs` UDF has been created):

1. The first (and only) argument to `abs` is `NULL` which is type `unknown`.
2. The `abs` candidates have parameter types `int2`, `int4`, `int8`, `float4`, `float8`, and `numeric`:

```
marcus=# SELECT proname, typname, typcategory, typispreferred FROM pg_proc JOIN pg_type ON proargtypes[0] = pg_type.oid WHERE proname = 'abs';
 proname | typname | typcategory | typispreferred
---------+---------+-------------+----------------
 abs     | int8    | N           | f
 abs     | int2    | N           | f
 abs     | int4    | N           | f
 abs     | float4  | N           | f
 abs     | float8  | N           | t
 abs     | numeric | N           | f
(6 rows)
```

3. The [category](https://www.postgresql.org/docs/current/catalog-pg-type.html#CATALOG-TYPCATEGORY-TABLE) for all of these types is numeric (N).
4. One of the candidates has the preferred type of the numeric category, `float8`. So, all other candidates are discarded.
5. Only a single candidate remains.

#### What about when we create a UDF named `abs` with a `TEXT` argument?

```sql
CREATE FUNCTION abs(t TEXT) RETURNS INT LANGUAGE SQL AS 'SELECT 3';
```

1. The first (and only) argument to `abs` is `NULL` which is type `unknown`.
2. The `abs` candidates have parameter types `int2`, `int4`, `int8`, `float4`, `float8`, `numeric`, and _now_ `text`.

```
marcus=# SELECT proname, typname, typcategory, typispreferred FROM pg_proc JOIN pg_type ON proargtypes[0] = pg_type.oid WHERE proname = 'abs';
 proname | typname | typcategory | typispreferred
---------+---------+-------------+----------------
 abs     | int8    | N           | f
 abs     | int2    | N           | f
 abs     | int4    | N           | f
 abs     | text    | S           | t
 abs     | float4  | N           | f
 abs     | float8  | N           | t
 abs     | numeric | N           | f
(7 rows)
```

3. Our UDF accepts a type of the string category (S), so we select that category.
4. Disgard the candidates which do not select the string category.
5. Only a single candidate, the UDF, remains.

We know that the UDF we created is the overload selected because we made it non-strict and it returns a non-NULL value when given NULL input.

```
marcus=# SELECT abs(NULL);
 abs
-----
   3
(1 row)
```

#### What about when we create another UDF named `abs` with a `CHAR` argument?

```sql
CREATE FUNCTION abs(t CHAR) RETURNS INT LANGUAGE SQL AS 'SELECT -1';
```

It's very similar to the last example up through (4). After that, we have two candidates remaining, the one that accepts `TEXT` and the one that accepts `CHAR`. We select the `TEXT` one because `TEXT` is the preferred type of the string category.

Proof that the `TEXT` overload is still selected, because the return value is `3`, not `-1`:

```
marcus=# SELECT abs(NULL);
 abs
-----
   3
(1 row)
```

#### Can `abs(NULL)` ever be ambiguous?

Sure. Let's drop our `abs` UDFs first and start with a clean slate.

```sql
DROP FUNCTION abs(TEXT);
DROP FUNCTION abs(CHAR);
```

One way to make `abs` ambiguous is to create a UDF named `abs` with parameter of a different type category:

```sql
CREATE FUNCTION abs(b BOOL) RETURNS INT LANGUAGE SQL AS 'SELECT 10';
```

```
marcus=# SELECT abs(NULL);
ERROR:  42725: function abs(unknown) is not unique
```

```sql
DROP FUNCTION abs(BOOL);
```

Another way is to create multiple UDFs with parameters of the string category that are not preferred (not `TEXT`):

```sql
CREATE FUNCTION abs(c CHAR) RETURNS INT LANGUAGE SQL AS 'SELECT 55';
CREATE FUNCTION abs(n NAME) RETURNS INT LANGUAGE SQL AS 'SELECT 66';
```

```
marcus=# SELECT abs(NULL);
ERROR:  42725: function abs(unknown) is not unique
```

#### So how do we fix this?

The long-term solution is to implement type categories and preferred types as Postgres has done, and use these type categories to fully implement this function overload resolution.

Ideally we can find a shorter term solution that both:

1. Removes the logic in function type-checking that disambiguates functions based on their strictness when presented with NULL (unknown) arguments.
2. Allows reasonable disambiguation of built-ins when presented with NULL types, so expressions like `abs(NULL)` do not error.

**Comment by rharding6373** [comment]
I tried to repro at head and it looks like the original issue has been fixed:

```
root@127.0.0.1:26257/defaultdb> CREATE TABLE t (a INT, b INT);                                                                                                       
                             -> INSERT INTO t VALUES (1, NULL);                                                                                                      
                             -> CREATE FUNCTION g (i INT) RETURNS INT CALLED ON NULL INPUT LANGUAGE SQL AS $$ SELECT a FROM t WHERE b IS NOT DISTINCT FROM i $$;     
                             -> SELECT g(NULL);                                                                                                                      
                             -> SELECT g(NULL::INT);                                                                                                                 
CREATE TABLE

Time: 6ms total (execution 5ms / network 1ms)

INSERT 0 1

Time: 9ms total (execution 9ms / network 0ms)

CREATE FUNCTION

Time: 34ms total (execution 26ms / network 8ms)

  g
-----
  1
(1 row)

Time: 3ms total (execution 3ms / network 0ms)

  g
-----
  1
(1 row)
```

**Comment by mgartner** [comment]
Huh... Maybe I completely mis-diagnosed this? Or we just happened to solve it with some other change. It looks like it's fixed on v23.1.

This example above looks fixed too:

```sql
CREATE FUNCTION f(i INT2) RETURNS INT STRICT LANGUAGE SQL AS 'SELECT 2';

CREATE FUNCTION f(i INT4) RETURNS INT STRICT LANGUAGE SQL AS 'SELECT 2';

-- This should error because f is ambiguous.
SELECT f(NULL);
```

It correctly returns an error.

I'm really curious what fixed this...

**Comment by rharding6373** [comment]
Do we need to find the fix and backport it? Both examples fail on 22.2.10 and the tip of 22.2.x.

**Comment by rharding6373** [comment]
I bisected the fix back to https://github.com/cockroachdb/cockroach/pull/98162, which was only[ backported to 23.1](https://github.com/cockroachdb/cockroach/pull/102188) since that's when we introduced record-returning UDFs.

What "fixed" this issue was the change in this line here:
https://github.com/rharding6373/cockroach/blob/4539072e48942b77721b1e3044235c1013266dc4/pkg/sql/sem/tree/type_check.go#L1143

This change prevents an early exit in function expression type checking when there is a UDF overload, NULL arguments that don't have a resolved type, and no called on null functions. This change was motivated by a bunch of testing with STRICT UDFs that was done as part of this PR.

So why does this fix a `CALLED ON NULL INPUT` function when we just said this only prevents early exit with ***no*** called on null input functions? I think that the changes in this PR may be masking another issue, which is that it appears that the UDF in the original example is being evaluated as a function that is ***not*** called on null input. 

The PR does fix the second example, though, since by bypassing the shortcut that returns null it allows us to hit the ambiguous function error instead.

I'm going to try to open a 22.2 backport PR with just the type check changes. I'm also going to investigate the `CALLED ON NULL INPUT` example a bit more, since it doesn't seem right.
